syntax = "proto3"; //声明protobuf的版本

package fixbug;   //声明了代码所在的包 （对于C++来说就是namespace)


//下面的选项，表示生成service服务类和rpc方法描述， 默认是不生成的
option cc_generic_services = true;


message ResultCode{  //定义返回的错误码
    int32 errcode = 1;
    bytes errmsg = 2;
}


//常用的数据类型： 数据、列表、映射表

//定义登录请求消息类型   name   pwd
message LoginRequest{
    bytes name = 1;   //等于1表示这是第一个参数，一般string的存储定义为bytes
    bytes pwd = 2;
   // map<int32,string> test = 3; //映射表类型
}

//定义登录响应消息类型
message LoginResponse{
    ResultCode result = 1;
    bool success = 3;
}

message GetFriendsListRequest{
    uint32 user_id = 1;
}

message User{
    bytes name = 1;
    uint32 age = 2;
    enum Sex{
        MAN = 0;
        WOMAN = 1;
    }
    Sex sex = 3;
}

message GetFriendsListResponse{
    ResultCode result = 1;
    repeated User friend_list = 2;  //定义了一个列表数据类型
}


//在protobuf里面怎么定义描述rpc方法的类型---service
service UserService{
    rpc Login(LoginRequest) returns(LoginResponse);
    rpc GetFriendsList(GetFriendsListRequest) returns(GetFriendsListResponse); 
}
